---
title: "Load PDX models and plot"
author: "Sara Gosline"
date: "7/26/2021"
output: 
  html_document:
    toc: true
    toc_depth: 2
---

This document focuses on loading the data and visualizing the PDX model data that we have.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(remotes)
if(!require('mpnstXenoModeling')){
  remotes::install_github('sgosline/mpnstXenoModeling')
  library(mpnstXenoModeling)
}

library(ggplot2)
library(dplyr)
loadPDXData()

```


## Load and plot PDX Data

PDX data is available on [this Synapse project page](https://www.synapse.org/#!Synapse:syn21984813/wiki/602362) that collates available PDX and tumor model data across the project.

For this project we create a function that loads and plots the data.

First let's summarize the data that we have currently. 

```{r PDX data summary,warning=FALSE}
#TODO: add in data summary table
```


## PDX drug experiments 

Now we can plot the individual experiments run on the PDXs to summarize.

```{r tumor data, message=FALSE,warning=FALSE}

plots <-drugData%>%
  group_by(Sample,batch,.drop=F)%>%
  group_map(~plotBySample(.x),.keep=TRUE)

pdf('allTumorVolume.pdf',height=8,width=12)

res=lapply(plots, print)
dev.off()
res=lapply(plots, print)

```

## Somatic variant data


Now we can look at the mutational data in more detail. What samples do we have somatic variants for? 

```{r mut data,echo=FALSE}
 mutMat<-varData%>%mutate(AD=as.numeric(AD))%>%
  dplyr::select(specimenID,Symbol,AD)%>%#,tranche))%>%
  subset(Symbol!="")%>%
  tidyr::pivot_wider(names_from=specimenID,values_from=AD,values_fn=list(AD=mean),
                     values_fill=0.0)%>%
    tibble::column_to_rownames('Symbol')

annotes <- varData%>%
  dplyr::select(Sample,specimenID,Age,Sex,MicroTissueQuality,Location,Size)%>%
  distinct()%>%
  tibble::column_to_rownames('specimenID')%>%
  mutate(Age=as.numeric(Age),Size=as.numeric(Size),MicroTissueQuality=unlist(MicroTissueQuality))
library(pheatmap)
 
  mutMat<-mutMat[,intersect(rownames(annotes),colnames(mutMat))]
  
 pheatmap(log10(0.01+mutMat),clustering_distance_cols = 'correlation',cellwidth = 10,annotation_col = annotes,labels_row = rep("",nrow(mutMat)),labels_col=rep("",ncol(mutMat)))
 
 pheatmap(log10(0.01+mutMat),clustering_distance_cols = 'correlation',cellwidth = 10,annotation_col = annotes,labels_row = rep("",nrow(mutMat)),labels_col=rep("",ncol(mutMat)),filename='allMutations.pdf')
```


There are so many one-off somatic variants, let's filter by those that exist in at least 3 samples and then add back TP53.
```{r main mutations,echo=FALSE}
topMuts=filter(varData,AD>0)%>%
  subset(!is.na(Symbol))%>%
  subset(Symbol!="")%>%
    group_by(Symbol)%>%
    summarize(nSamps=n_distinct(individualID))%>%
    filter(nSamps>2)%>%
    dplyr::select(Symbol)

topMuts<-union("TP53",topMuts$Symbol)

 pheatmap(log10(0.01+mutMat[topMuts,rownames(annotes)]),cellwidth = 10,cellheight=10,annotation_col = annotes,clustering_method='ward.D2')
 
 pheatmap(log10(0.01+mutMat[topMuts,]),cellwidth = 10,cellheight=10,annotation_col = annotes,filename='recMutations.pdf',clustering_method='ward.D2')

```

## Gene expression data

We can also plot the gene expression across highly variable genes.
```{r mutated gene expression,echo=FALSE,warning=FALSE}
 rnamat<-rnaSeq%>%
  subset(!is.na(Symbol))%>%
  subset(!is.na(totalCounts))%>%
  dplyr::select(-c(individualID,species,sex,zScore,experimentalCondition))%>%
  tidyr::pivot_wider(names_from=specimenID,values_from=totalCounts,
                     values_fn=list(totalCounts=mean),values_fill=0.0)%>%
    tibble::column_to_rownames('Symbol')

rownames(rnaSeq)<-c()
 annotes<-rnaSeq%>%
      dplyr::select(specimenID,individualID,sex)%>%
   distinct()%>%
    tibble::column_to_rownames('specimenID')

 #topMuts<-setdiff(topMuts,'TTN')
pheatmap(log10(0.01+rnamat[intersect(topMuts,rownames(rnamat)),]),
          clustering_distance_cols = 'correlation',
          cellwidth = 10,
       #  annotation_col = annotes,
         cellheight=10,
         labels_col=rep("",ncol(mutMat)))

 pheatmap(log10(0.01+rnamat[intersect(topMuts,rownames(rnamat)),]),
          clustering_distance_cols = 'correlation',
          cellwidth = 10,
        # annotation_col = annotes,
         cellheight=10,
         labels_col=rep("",ncol(mutMat)),filename='countsOfMutatedGenes.pdf')
 
 
```

## Microtissue data
Lastly we have the microtissue data that we we also have drug treatment for.
```{r plot MT data, warning=FALSE}

#TODO: summarize MT data as well
```

This is the data we have so far!