---
title: "Expression correlation analysis"
author: "Sara gosline"
date: "8/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

require(remotes)
if(!require(mpnstXenoModeling)){
  remotes::install_github('sgosline/mpnstXenoModeling')
  library(mpnstXenoModeling)
}


```

Here we bring in the RNASeq data and the `leapR` package to measure and identify correlated pathways across the samples.


## Load data from synapse query


```{r loadData,warning=FALSE}
#this function simply loads all the data into memory

syn<-reticulate::import('synapseclient')$login()
data.tab<-syn$tableQuery('select * from syn24215021')$asDataFrame()
rnaSeq<<-mpnstXenoModeling::dataFromSynTable(data.tab,syn,'RNASeq')#%>%



```


## Filter genes by low expression


```{r filtering, warning=FALSE}


```

## First let's plot the most variable fully expressed genes

We want to plot the variable genes in a heatmap and also the PCA


```{r Pca plot}
plotAllData <- function(dat.table, expand=0.01, alpha=0.1, ...) {
  library(ggfortify)
  met <- dat.table %>%
    dplyr::select(Sample, MicroTissueQuality, Clinical.Status, Sex) %>%
    distinct()%>%
    mutate(MicroTissueQuality=unlist(MicroTissueQuality))%>%
    mutate(Clinical.Status=unlist(Clinical.Status))

    
  mat <- dat.table %>% dplyr::select(TXID,NumReads,Sample) %>%
    distinct() %>%
    mutate(Reads=as.numeric(NumReads)) %>%
    select(-NumReads)%>%
    tidyr::pivot_wider(names_from='Sample', values_from='Reads',
                       values_fn=list(Reads=mean),
                       values_fill=list(Reads=0)) %>%
    tibble::remove_rownames() %>%
    tibble::column_to_rownames('TXID')
 # mat <- mat[complete.cases(mat),]
  x <- prcomp(t(mat))$x
  x <- as.data.frame(x)
  x$Sample <- rownames(x)
  ggdata <- inner_join(x, met, by="Sample")
  
  library(ggforce)
  ggplot(ggdata, aes_string(x="PC1", y="PC2", color = 'MicroTissueQuality',shape='Clinical.Status')) +
    geom_point(size=2.5)# +
    #ggforce::geom_mark_ellipse( alpha=alpha, expand=expand)
 
}

plotAllData(master.table)
```


```{r most variable}


by.gene <- master.table%>%
  group_by(Symbol,Sample)%>%
  summarize(totalCounts=sum(NumReads))

pat.var <-by.gene%>%group_by(Symbol)%>%
  summarize(geneVar=var(totalCounts))%>%
  arrange(desc(geneVar))

symbs<-pat.var$Symbol[1:100]


```

## Correlate all genes across Samples

```{r leapr cor}

```
