---
title: "Expression correlation analysis"
author: "Sara gosline"
date: "8/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

require(remotes)
if(!require(mpnstXenoModeling)){
  remotes::install_github('sgosline/mpnstXenoModeling')
  library(mpnstXenoModeling)
}

if(!require(leapr)){
  remotes::install_github('biodataganache/leapR')
  library(leapr)
}



```

Here we bring in the RNASeq data and the `leapR` package to measure and identify correlated pathways across the samples.


## Load data from synapse query


```{r loadData,warning=FALSE}
#this function simply loads all the data into memory

syn<-reticulate::import('synapseclient')$login()
data.tab<-syn$tableQuery('select * from syn24215021')$asDataFrame()
rnaSeq<<-mpnstXenoModeling::dataFromSynTable(data.tab,syn,'RNASeq')#%>%
master.table<-rnaSeq
```


## Normalize counts


```{r filtering, warning=FALSE}

normalize_counts_deseq2<-function(long.table){
  library(DESeq2)
  require(dplyr)
  require(tidyr)
  count.mat<-master.table%>%
    dplyr::select(TXID,NumReads,Sample)%>%
    pivot_wider(values_from=NumReads,names_from=Sample)%>%
    tibble::column_to_rownames('TXID')
  met.mat<-master.table%>%
    dplyr::select(Sample,Age,MicroTissueQuality,Sex,Clinical.Status)%>%
    distinct()%>%
    tibble::column_to_rownames('Sample')
  
  dds <- DESeqDataSetFromMatrix(countData = count.mat, colData = met.mat)

}

```

## First let's plot the most variable fully expressed genes

We want to plot the variable genes in a heatmap and also the PCA


```{r Pca plot}
plotAllExpData <- function(dat.table, expand=0.01, alpha=0.1, ...) {
  library(ggfortify)
  met <- dat.table %>%
    dplyr::select(Sample, MicroTissueQuality, Clinical.Status, Sex) %>%
    distinct()%>%
    mutate(MicroTissueQuality=unlist(MicroTissueQuality))%>%
    mutate(Clinical.Status=unlist(Clinical.Status))

    
  mat <- dat.table %>% dplyr::select(TXID,NumReads,Sample) %>%
    distinct() %>%
    mutate(Reads=as.numeric(NumReads)) %>%
    dplyr::select(-NumReads)%>%
    tidyr::pivot_wider(names_from='Sample', values_from='Reads',
                       values_fn=list(Reads=mean),
                       values_fill=list(Reads=0)) %>%
    tibble::remove_rownames() %>%
    tibble::column_to_rownames('TXID')
 # mat <- mat[complete.cases(mat),]
  x <- prcomp(t(mat))$x
  x <- as.data.frame(x)
  x$Sample <- rownames(x)
  ggdata <- inner_join(x, met, by="Sample")
  
  library(ggforce)
  ggplot(ggdata, aes_string(x="PC1", y="PC2", color = 'MicroTissueQuality',shape='Clinical.Status')) +
    geom_point(size=2.5)# +
    #ggforce::geom_mark_ellipse( alpha=alpha, expand=expand)
 return(mat)
}

data.mat<-plotAllExpData(master.table)
```


```{r most variable}
library(tidyr)

by.gene <- master.table%>%
  group_by(Symbol,Sample)%>%
  summarize(totalCounts=sum(NumReads))

pat.var <-by.gene%>%group_by(Symbol)%>%
  summarize(geneVar=var(totalCounts))%>%
  arrange(desc(geneVar))


pat.annote <- master.table %>%
    dplyr::select(Sample, MicroTissueQuality, Clinical.Status, Sex) %>%
    distinct()%>%
    mutate(MicroTissueQuality=unlist(MicroTissueQuality))%>%
    mutate(Clinical.Status=unlist(Clinical.Status))%>%
  tibble::column_to_rownames('Sample')


symbs<-pat.var$Symbol[1:100]

count.mat<-by.gene%>%
  dplyr::select(totalCounts,Symbol,Sample)%>%
  pivot_wider(values_from=totalCounts,names_from=Sample,values_fn=list(totalCounts=mean))%>%
  tibble::column_to_rownames('Symbol')

library(pheatmap)
pheatmap(log10(0.01+count.mat[symbs,]),cellwidth = 10,cellheight=10,annotation_col = pat.annote,filename='mostVarTrans.pdf')
pheatmap(log10(0.01+count.mat[symbs,]),cellwidth = 10,cellheight=10,annotation_col = pat.annote)
```

## Correlate all genes across Samples

```{r leapr cor}

##now let's look at the pathways that are correlated across all
library(leapr)
#data(ncipid)
data("krbpaths")
idx.kegg <- grepl("^KEGG_", krbpaths$names)
names.kegg <- krbpaths$names[idx.kegg]
names.kegg <- sub("KEGG_", "", names.kegg)
names.kegg <- gsub("_", " ", names.kegg)
names.kegg <- sapply(names.kegg, function(y) paste(strwrap(y, 45), 
                                          collapse = "\n"), 
                     USE.NAMES = FALSE)
desc.kegg <- krbpaths$desc[idx.kegg]
sizes.kegg <- krbpaths$sizes[idx.kegg]
Max <- max(sizes.kegg)
matrix.kegg <- krbpaths$matrix[idx.kegg, 1:Max]
keggpaths <- list(names = names.kegg,
                 desc = desc.kegg,
                 sizes = sizes.kegg,
                 matrix = matrix.kegg)
#term.2.gene <- as.data.frame(keggpaths$matrix) %>%
#  mutate(term = keggpaths$names) %>%
#  pivot_longer(!term, names_to = "Column", values_to = "gene") %>%
#  filter(!(gene == "null")) %>%
#  select(term, gene)
#term.2.name <- data.frame(term = keggpaths$names, name = keggpaths$names)
  
plotCorrelationEnrichment(as.matrix(data.mat), keggpaths, fdr.cutoff = 0.05, 
                                      corr.cutoff = 0.1, prefix='allSamps', width = 11, 
                                      height = 8.5, order.by = "Ingroup mean", 
                                      clean.names = FALSE, pathway.plot.size = 3)


```
