---
title: "Fig2 MT Drug Sensitivity Expression Correlation"
author: "Sara Gosline"
date: "9/13/2021"
output:  
  html_document:
    toc: true
    toc_depth: 2
---

This markdown assesses gene expression changes the correlate with drug sensitivity.

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)

require(remotes)
if(!require(mpnstXenoModeling)){
  remotes::install_github('sgosline/mpnstXenoModeling')
  library(mpnstXenoModeling)
}

if(!require('leapr')){
  remotes::install_github('biodataganache/leapR')
  library(leapr)
}

if(!require('DT')){
  install.packages("DT")
  library(leapr)
}

library(dplyr)

```


## Load drug data
```{r mutation data,message=FALSE,warning=FALSE}
mpnstXenoModeling::loadPDXData()

```


## Drug Plots

We can now evaluate drugs that behave differently across samples. We are primarily focused on four samples and four drugs. However, we anticipate pooling the results for each drug to carry out differential expression.


### Microtissue plots

Which MT drugs do we focus on?

```{r MT drug plots, message=FALSE,warning=FALSE}
syn = synapseLogin()
mtStats <- syn$tableQuery("select * from syn26136282")$asDataFrame()%>%
  mutate(maxKilling = 1-(MinViability/MaxViability))

good.mt <- subset(clin.tab,MicroTissueQuality%in%c('Good','Robust','Usable'))%>%distinct()

print(paste('Looking at ',nrow(good.mt),'micro tisues'))

mtStats <- mtStats%>%
  subset(CellLine%in%good.mt$Sample)

minDrugs <- mtStats%>%
   mutate(hasRNASeq=(CellLine%in%rnaSeq$Sample))%>%
  group_by(Drug)%>%
  summarise(samps=n_distinct(CellLine),samplesWithRNA=count(hasRNASeq==TRUE))%>%
  subset(samps>3)

DT::datatable(minDrugs)

```

We have many more data points to correlate with the MT data.

```{r mt stat plots}

combos<-unique(minDrugs$Drug[grep(';',minDrugs$Drug)])
singles<-setdiff(minDrugs$Drug,combos)

drug.plots <- lapply(singles,function(dname){
  mtStats%>%
    subset(Drug==dname)%>%
    ggplot(aes(x=maxKilling,y=auc))+geom_point()+ggtitle(dname)
})

ggplot(subset(mtStats,Drug%in%singles),aes(x=Drug,y=auc,fill=CellLine))+
    geom_bar(stat='identity',position='dodge')+scale_x_discrete(guide = guide_axis(angle=90))
ggplot(subset(mtStats,Drug%in%singles),aes(x=Drug,y=maxKilling,fill=CellLine))+
    geom_bar(stat='identity',position='dodge')+scale_x_discrete(guide = guide_axis(angle=90))

cowplot::plot_grid(plotlist=drug.plots,nrow=7)

drug.plots <- lapply(combos,function(dname){
  mtStats%>%
    subset(Drug==dname)%>%
    ggplot(aes(x=maxKilling,y=auc))+geom_point()+ggtitle(dname)
})


cowplot::plot_grid(plotlist=drug.plots,nrow=3)

ggplot(subset(mtStats,Drug%in%combos),aes(x=Drug,y=auc,fill=CellLine))+
    geom_bar(stat='identity',position='dodge')+scale_x_discrete(guide = guide_axis(angle=90))

ggplot(subset(mtStats,Drug%in%combos),aes(x=Drug,y=maxKilling,fill=CellLine))+
    geom_bar(stat='identity',position='dodge')+scale_x_discrete(guide = guide_axis(angle=90))


```
We apparently have a lot of variability here, which is a good thing. We can now distinguish cells that are sensitive (AUC<0.6) compared to those that are not (AUC>0.6) to evaluate differential expression.

```{r MT stats}

auc.thresh=0.6

sens.class <- mtStats%>%
  mutate(sens=(auc<auc.thresh))

diff.drugs=sens.class%>%
  group_by(Drug)%>%
  summarize(res=count(!sens),sens=count(sens))%>%
  subset(sens>1)%>%
  subset(res>1)

diff.drugs

p<-ggplot(subset(mtStats,Drug%in%diff.drugs$Drug),aes(x=Drug,y=auc,fill=CellLine))+
    geom_bar(stat='identity',position='dodge')+scale_x_discrete(guide = guide_axis(angle=90))
#ggsave(paste0('drugsWith2SensAt',auc.thresh,'.png'),p)
p
```


## RNA Expression


```{r rnaseq}


    library(ensembldb)
  
  database <- EnsDb.Hsapiens.v86
  pmap <- ensembldb::select(database, keys=list(GeneIdFilter(rnaSeq$GENE), TxBiotypeFilter("protein_coding")),
      columns = c("GENENAME"))%>%
    dplyr::rename(GENE='GENEID')%>%
    right_join(rnaSeq)%>%
    subset(TXBIOTYPE=='protein_coding')%>%
    dplyr::select(GENE,GENENAME,GENEID)%>%distinct()%>%
    tibble::column_to_rownames('GENEID')
  
  dds<-rnaSeq%>%subset(GENE%in%pmap$GENE)%>%
    mpnstXenoModeling::deseq2NormFilter()


```

### Genes correlated with PDX hits

### Genes correlated with MT hits

The goal for this is to subset the patients by whether or not they are sensitive to different drugs then compare the expression for each. There are `$nrow(diff.drugs)`drugs that exhibit variable behavior with an AUC less than `$auc.thresh`. 

```{r diffex tests}

##let's rename the samples a bit? 

all.res<-lapply(diff.drugs$Drug,function(x){
  sens.pats <- subset(sens.class,Drug==x)%>%subset(sens==TRUE)
  res.pats <- subset(sens.class,Drug==x)%>%subset(sens==FALSE)

  ##add col data to dds
    
  drug.res <- mpnstXenoModeling::ds2FactorDE(dds, ids1=sens.pats$CellLine,
                                    ids2=res.pats$CellLine,name=x)
  
  mpnstXenoModeling::plotTopGenesHeatmap(drug.res, dds,
                                         pmap,x,patients=c(sens.pats$CellLine,res.pats$CellLine),
                                         adjpval=0.01, upload=FALSE, parentID='syn25323411')
  
  #pmap[rownames(subset(drug.res,padj<0.01)),'GENENAME']%>%doRegularGo(prefix=x)
  
  genes.with.values<-cbind(drug.res,pmap[rownames(drug.res),])%>%
  dplyr::select(Gene='GENENAME',value='log2FoldChange')
                   
  res2=doGSEA(genes.with.values,prot.univ=NULL,prefix=x,useEns=FALSE)
  drug.res%>%mutate(Drug=x)
  })

full.res<-do.call(rbind,all.res)
                                   
```


Now we have the enrichment for each of the drugs. 
```

### Check for overlap